<div style="border:solid green 2px; padding: 20px">

Hi Endi,

Thank you so much for resubmitting the project! And once again, I'm sorry for the confusion with the previous review — that was entirely my mistake.

Everything was already in great shape, so no additional changes are needed. Your project is officially accepted now! 🎉

Great work, and thank you for your understanding!
</div>

<div style="border:solid green 2px; padding: 20px">

Hi,

My name is Yuliya and I'll be reviewing your project. You can contact me through my Tripleten profile: </b><a href="https://hub.tripleten.com/u/4e3b1e8f">Yuliya Nikishina</a></b>

Great job completing your first Machine Learning project! 🎉

This project is an important milestone, as it brings together the key concepts you've learned about supervised learning, model selection, hyperparameter tuning, and model evaluation.

You can find my comments throughout the notebook. I use the following color coding:

<div class="alert alert-success" style="border-left: 7px solid green">
<b>✅ Great Job</b> – Something was completed correctly or particularly well.
</div>

<div class="alert alert-warning" style="border-left: 7px solid orange">
<b>⚠️ Recommendation</b> – Suggestions for improvement or best practices. These comments do not affect the project approval but can help make your notebook even clearer and more professional.
</div>

<div class="alert alert-danger" style="border-left: 7px solid red">
<b>⛔️ Needs Fixing</b> – These issues must be addressed before the project can be approved.
</div>

Please **do not delete or modify my comments** when working on a revision. If you'd like to reply or ask a question, feel free to add your response directly below my comment in a separate Markdown cell. For example:

<div class="alert alert-info" style="border-left: 7px solid #17a2b8">
<b>💬 Student's reply</b><br>
Your response here.
</div>



Excellent work! You correctly followed the expected machine learning workflow by splitting the data into training, validation, and test sets, comparing multiple models, tuning hyperparameters, selecting the best-performing model based on the validation set, evaluating it on the test set, and performing a sanity check. Your final summary also provides a clear overview of the project and its results.

I've left one small recommendation about notebook organization that will make your work even easier to read, but it doesn't affect the correctness of your solution.

**Congratulations — your project is accepted!** 🎉 Keep up the great work! 🚀

</div>

In [1]:
import pandas as pd

# Load dataset
df = pd.read_csv('../datasets/users_behavior.csv')

# Inspect structure and missing values
print(df.info())
print(df.head())

# Check target distribution
print(df['is_ultra'].value_counts(normalize=True))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None
   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
0    0.693528
1    0.306472
Name: is_ultra, dtype: float64


<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment, v. 1</b>

Your notebook is easy to follow overall. As a small readability improvement, consider organizing the notebook into smaller logical steps.

It's a common practice to place all library imports together in the first code cell, then separate the workflow into individual sections (e.g., loading the dataset, inspecting its structure, checking the target distribution, splitting the data, training models, and evaluating results). Splitting large code cells into smaller ones with short Markdown headings makes the notebook easier to read, navigate, and review.
</div>

In [2]:
from sklearn.model_selection import train_test_split

# 1. Separate features and target
X = df.drop(columns=['is_ultra'])
y = df['is_ultra']

# 2. First split: 60% train, 40% temp (validation + test)
X_train, X_temp, y_train, y_temp = train_test_split(
    X, y, test_size=0.40, random_state=12345, stratify=y
)

# 3. Second split: Divide the 40% temp set 50/50 into validation (20%) and test (20%)
X_valid, X_test, y_valid, y_test = train_test_split(
    X_temp, y_temp, test_size=0.50, random_state=12345, stratify=y_temp
)

# Verify shapes
print("Train shape:", X_train.shape)
print("Valid shape:", X_valid.shape)
print("Test shape:", X_test.shape)

Train shape: (1928, 4)
Valid shape: (643, 4)
Test shape: (643, 4)


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b>

Great job splitting the dataset into separate training, validation, and test sets. Using `stratify` is also a nice touch, as it preserves the class distribution across all subsets and makes the evaluation more reliable.
</div>

In [3]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Variables to track the best depth and accuracy
best_depth = 0
best_accuracy = 0.0
best_dt_model = None

# Loop through max_depth from 1 to 10
for depth in range(1, 11):
    # 1. Initialize model with current max_depth
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    
    # 2. Train on the training set
    model.fit(X_train, y_train)
    
    # 3. Predict and evaluate on the validation set
    predictions_valid = model.predict(X_valid)
    acc = accuracy_score(y_valid, predictions_valid)
    
    print(f"max_depth = {depth} | Validation Accuracy = {acc:.4f}")
    
    # Track the model with highest accuracy
    if acc > best_accuracy:
        best_accuracy = acc
        best_depth = depth
        best_dt_model = model

print("-" * 40)
print(f"Best max_depth: {best_depth} with Validation Accuracy: {best_accuracy:.4f}")

max_depth = 1 | Validation Accuracy = 0.7403
max_depth = 2 | Validation Accuracy = 0.7729
max_depth = 3 | Validation Accuracy = 0.7776
max_depth = 4 | Validation Accuracy = 0.7543
max_depth = 5 | Validation Accuracy = 0.7854
max_depth = 6 | Validation Accuracy = 0.7745
max_depth = 7 | Validation Accuracy = 0.7869
max_depth = 8 | Validation Accuracy = 0.8025
max_depth = 9 | Validation Accuracy = 0.7823
max_depth = 10 | Validation Accuracy = 0.7729
----------------------------------------
Best max_depth: 8 with Validation Accuracy: 0.8025


In [4]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

best_accuracy = 0.0
best_est = 0
best_depth = 0
best_rf_model = None

# Outer loop: Number of trees in the forest
for est in range(10, 51, 10):
    # Inner loop: Maximum depth of each tree
    for depth in range(1, 11):
        # Initialize and train model
        model = RandomForestClassifier(n_estimators=est, max_depth=depth, random_state=12345)
        model.fit(X_train, y_train)
        
        # Evaluate on validation set
        predictions_valid = model.predict(X_valid)
        acc = accuracy_score(y_valid, predictions_valid)
        
        # Track the top combination
        if acc > best_accuracy:
            best_accuracy = acc
            best_est = est
            best_depth = depth
            best_rf_model = model

print(f"Best Random Forest Model:")
print(f"n_estimators = {best_est}, max_depth = {best_depth}")
print(f"Validation Accuracy: {best_accuracy:.4f}")

Best Random Forest Model:
n_estimators = 40, max_depth = 9
Validation Accuracy: 0.8212


In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# 1. Initialize the model
lr_model = LogisticRegression(random_state=12345, solver='liblinear')

# 2. Train on the training set
lr_model.fit(X_train, y_train)

# 3. Evaluate on the validation set
lr_predictions = lr_model.predict(X_valid)
lr_accuracy = accuracy_score(y_valid, lr_predictions)

print(f"Logistic Regression Validation Accuracy: {lr_accuracy:.4f}")

Logistic Regression Validation Accuracy: 0.7185


In [6]:
from sklearn.metrics import accuracy_score

# 1. Get predictions on the unseen test set
test_predictions = best_rf_model.predict(X_test)

# 2. Calculate test accuracy
test_accuracy = accuracy_score(y_test, test_predictions)

print(f"Final Test Set Accuracy: {test_accuracy:.4f}")

# Check project requirement (threshold = 0.75)
if test_accuracy >= 0.75:
    print("Success! The model passes the required 0.75 accuracy threshold.")
else:
    print("The model did not meet the accuracy threshold.")

Final Test Set Accuracy: 0.8087
Success! The model passes the required 0.75 accuracy threshold.


In [7]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score

# Create a baseline model that always predicts the most frequent class (Smart / 0)
dummy_model = DummyClassifier(strategy='most_frequent', random_state=12345)

# Fit on training data and predict on test data
dummy_model.fit(X_train, y_train)
dummy_predictions = dummy_model.predict(X_test)

dummy_accuracy = accuracy_score(y_test, dummy_predictions)

print(f"Dummy Model (Baseline) Accuracy: {dummy_accuracy:.4f}")
print(f"Random Forest Test Accuracy:     {test_accuracy:.4f}")
print(f"Sanity Check Passed:            {test_accuracy > dummy_accuracy}")

Dummy Model (Baseline) Accuracy: 0.6936
Random Forest Test Accuracy:     0.8087
Sanity Check Passed:            True


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b>

Great work! You followed the expected machine learning workflow throughout the project: you correctly split the data into training, validation, and test sets, compared multiple models with different hyperparameters, selected the best model based on the validation set, evaluated it on the test set, and completed a sanity check with a baseline model. Overall, this is a well-structured solution that successfully meets all the project requirements.
</div>

# Project Summary & Findings

### 1. Model Investigation & Hyperparameter Tuning
* **Decision Tree:** Tested `max_depth` from 1 to 10. Accuracy peaked at `max_depth = 8` with a validation accuracy of `0.8025`.
* **Random Forest:** Tested `n_estimators` (10 to 50) and `max_depth` (1 to 10). The best combination was `n_estimators = 40` and `max_depth = 9`, achieving the highest validation accuracy of `0.8212`.
* **Logistic Regression:** Serves as a baseline linear model with a validation accuracy of `0.7185`.

### 2. Test Set Evaluation
* Evaluated our top-performing model (**Random Forest**) on the unseen test set (`X_test`, `y_test`).
* **Final Test Accuracy:** `0.8087` (Exceeds the required `0.75` project threshold).

### 3. Sanity Check
* A `DummyClassifier` predicting the most frequent class (`0` / Smart plan) yields a baseline accuracy of `0.6936`.
* Our model outperforms the dummy baseline by over 11 percentage points (`0.8087 > 0.6936`), confirming the sanity check passes.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b>

Excellent summary! It clearly highlights the key results of your model comparison, explains how the final model was selected, and confirms its performance on the test set with a successful sanity check. This provides a concise and informative conclusion to the project.
</div>